### ISSUE TITLE
Remove remaining notions of the `clear` password field

### ISSUE THREAD

**Issue by rimas-kudelis** [avleen/vexim2#37]
Apparently, in #23 not all notions of the `clear` field were removed. This may have lead us to #35 and #36, where patches have been proposed which were supposed "fix" non-handling of that field in the code (basically by undoing #14).
We should get rid of these remnants.
@Udera, would you perhaps like to do this?

**Comment by rimas-kudelis** [comment]
The files that seem to be still affected are under `docs` and `setup`:
* `docs/configure`
* `docs/database-spec.txt`
* `docs/clients/mysql-popper.conf`
* `setup/create_db.pl`
* `setup/mysql.sql`
* `setup/pgsql.sql`

**Comment by Udera** [comment]
I can have a look into it, removing it shouldn't be a problem.
I was wondering what happens with users who want to update their setup? This requires more migration steps (some might want to change the hashing algorithm as md5 is probably not state of the art).

**Comment by rimas-kudelis** [comment]
AFAIK, Exim4U has written some migration scripts, we should see if we could reuse that.

But frankly speaking, I'm not sure the current Vexim codebase is worth improving. Coding standards and practices have improved a lot since it was written, so it might be (and I think it is) better to just rewrite it from scratch, using a contemporary approach (modern PHP5 framework, separation of php code from templates etc.). Ideally, the framework would have built-in support for migrations (and migration generations), so we could concentrate on writing and improving business logic instead of dealing with a constantly increasing amount of database versions we have to migrate from.

In #33, @Z3po suggested something similar: his idea is that we should ship this release and start working on a new one. I think, this could be our chance to start from the ground up. Would you be interested in that?

**Comment by Udera** [comment]
You are probably right with a new start and your ideas sound good. I would be happy to help out but I have to say that I'm not a software developer.

**Comment by Z3po** [comment]
I'd vote for that, too.
Let's build up a "clean" stable release (probably with the clear-field back in.
I'm no web developer either but would be happy to add some lines of code when having time for it...

**Comment by rimas-kudelis** [comment]
I don't think we really want the `clear` field back. Unless I misunderstood what you meant. :)

**Comment by Z3po** [comment]
So we have 2 issues here:
* nobody likes to have the clear field. I agree with that, but:
* we've introduced the clear field and don't have any migration scripts ready for removing it within the already existing mysql tables...

So I see the following approaches:
* freeze with the clear removed and have "broken" database setups
* freeze with the clear fiels being supported/used and clean it up in database upgrade scripts

I have no Problems with keeping it removed but I know how to alter the tables....
we could also just remove it and check for existance in later migration scripts to finally clean it up succesfully...should work, too.
I'm not sure about the "best" way here.

**Comment by rimas-kudelis** [comment]
If you set up the current state of Vexim now, I think the `clear` field would exist in the database, but it would never be filled in by anything. So basically, it would be non-functional.

I guess we could remove it completely by writing migration scripts, release Vexim v. 2.2 (or whatever) and be done with this old codebase.Would that be a good way?

**Comment by Z3po** [comment]
agree.

**Comment by Udera** [comment]
To make it work for everybody, we could copy the content of the clear-fild (clear-text password) into the crypt field. Then we can delete the clear field and set the cryptscheme to clear https://github.com/avleen/vexim2/blob/master/vexim/config/variables.php#L35.
Advantage: this approach works for everybody, in a further step the clear-password can then be converted to any hash-method (md5, sha1, sha2, ...).

**Comment by Udera** [comment]
Ok, I prepared a pull request https://github.com/Udera/vexim2/commit/cbc2f19482980acf99e5e36ba8f2eb94332c94b0
But I still have to put a switch in the installer-script, if the password is clear or hashed.

**Comment by rimas-kudelis** [comment]
Re installer script, are you talking about create_db.pl? If so, yeah, you can probably add a switch.

BTW, I've downloaded all versions of Vexim I could find to a VM. I'm thinking about probably attempting to write a universal migration script for Vexim4, if that ever happens. I haven't yet looked at how often the DB schema changed, perhaps there isn't much to write really...

Regarding Exim not supporting some hash algos: I think this should be mentioned, yes, but I don't think we absolutely MUST ship example configs for all other means of authentication. If there are some bits we don't know, we can simply say that we don't know. It's up to our user to choose a right algorithm wisely.

Regarding Courier: incidently, that's what I'm using at the moment. Here are working authenticators against courier-authdaemon (I'm not really sure where I have them from, could be Debian; also bear in mind that some of these uppercase macros are probably defined elsewhere in Debian files):
```
# Authenticate against courier authdaemon

# This is now the (working!) example from
# http://www.exim.org/eximwiki/FAQ/Policy_controls/Q0730
# Possible pitfall: access rights on /var/run/courier/authdaemon/socket.
 plain_courier_authdaemon:
   driver = plaintext
   public_name = PLAIN
   server_condition = \
     ${extract {ADDRESS} \
               {${readsocket{/var/run/courier/authdaemon/socket} \
               {AUTH ${strlen:exim\nlogin\n$auth2\n$auth3\n}\nexim\nlogin\n$auth2\n$auth3\n} }} \
               {yes} \
               fail}
   server_set_id = $auth2
   .ifndef AUTH_SERVER_ALLOW_NOTLS_PASSWORDS
#  server_advertise_condition = ${if or {{eq{$tls_in_cipher}{}}{eq{$host_address}{$received_ip_address}}}{}{*}}
#  server_advertise_condition = ${if eq{$tls_in_cipher}{}{}{*}}
   server_advertise_condition = ${if or{\
       {!eq{$tls_cipher}{}}\
       {match_ip {$sender_host_address}{@[]}}\
        }\
        {*}{}}
   .endif

 login_courier_authdaemon:
   driver = plaintext
   public_name = LOGIN
   server_prompts = Username:: : Password::
   server_condition = \
     ${extract {ADDRESS} \
               {${readsocket{/var/run/courier/authdaemon/socket} \
               {AUTH ${strlen:exim\nlogin\n$auth1\n$auth2\n}\nexim\nlogin\n$auth1\n$auth2\n} }} \
               {yes} \
               fail}
   server_set_id = $auth1
   .ifndef AUTH_SERVER_ALLOW_NOTLS_PASSWORDS
#  server_advertise_condition = ${if or {{eq{$tls_in_cipher}{}}{eq{$host_address}{$received_ip_address}}}{}{*}}
#  server_advertise_condition = ${if eq{$tls_in_cipher}{}{}{*}}
   server_advertise_condition = ${if or{\
       {!eq{$tls_cipher}{}}\
       {match_ip {$sender_host_address}{@[]}}\
        }\
        {*}{}}
   .endif
```

**Comment by rimas-kudelis** [comment]
By the way, regarding the clear field: from my understanding, it should work even without data conversion, no?

**Comment by Udera** [comment]
It depends, if someone used the clear-field in the exim-authenticator and the imap-authenticator.

**Comment by Udera** [comment]
You convert the database and create/recreate the salted hashes, but you always need to modifier the authenticator-section of exim and imap manually, and the variables of vexim.

**Comment by Udera** [comment]
I made a pull request: https://github.com/avleen/vexim2/pull/53

I will have a look into the perl script and the switch (I'd like to add SHA512). Then we need to check exim, dovecot, and courier configuration.

**Comment by rimas-kudelis** [comment]
Fixed in #53. Thanks @Udera !
